src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java

```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            return comparison.areEqual() ? null : "SamComparison fails";
+        }
+    }
+
+    /**
+     * Compares SAM/BAM files in a stringent way but not by byte identity (allow reorder of attributes)
+     * Comparing by MD5s is too strict and comparing by SamComparison is too lenient. So we need this method.
+     * Returns null if equal or message string if not equal.
+     */
+    public static String samsEqualStringent(final File sam1, final File sam2, final ValidationStringency validation, final File reference) throws IOException {
+        //Three phases:
+        //1:  if MD5s are same, then we're equal. Quick check
+        final String fileMD5_1 = Utils.calculateFileMD5(sam1);
+        final String fileMD5_2 = Utils.calculateFileMD5(sam2);
+        if (fileMD5_1.equals(fileMD5_2)){
+            return null;
+        }
+
+        //2:  SamComparison says no, then we're not equal
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)){
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            if (! comparison.areHeadersEqual()){  //For headers, we want identity (at least for now)
+                return "Headers different";
+            }
+            if (!comparison.areEqual()) {
+                return "SamComparison failed";
+            }
+        }
+
+        //At this point we know that the files are not byte-wise identical, but are equal according to SamComparison and their headers are equal
+        //So we iterate over reads and compare them one by one.
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)) {
+            final SAMRecordIterator it1 = reader1.iterator();
+            final SAMRecordIterator it2 = reader2.iterator();
+            while (it1.hasNext() && it2.hasNext()) {
+                final SAMRecord read1 = it1.next();
+                final SAMRecord read2 = it2.next();
+                final String eqMessage = readsEqualAllowAddingAttributes(read1, read2);
+                if (eqMessage != null){
+                    return eqMessage;
+                }
+            }
+            if (it1.hasNext() || it2.hasNext()) {
+                //at least one has no more records (because the while loop is done) and at least one does have records. So we're not equal.
+                return "Not the same number of reads";
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Compares the reads but ignores order or attributes.
+     * Also allows read2 to have a superset of attributes of read1.
+     */
+    private static String readsEqualAllowAddingAttributes(final SAMRecord read1, final SAMRecord read2) {
```

Comment:
Extract the attribute comparison code out as a separate helper method to break this long-ish method up.